sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamUnnestRel.java

```diff
@@ -111,31 +112,41 @@ public RelWriter explainTerms(RelWriter pw) {
       Schema joinedSchema = CalciteUtils.toSchema(getRowType());
 
       return outer
-          .apply(ParDo.of(new UnnestFn(joinedSchema, unnestIndex)))
+          .apply(ParDo.of(new UnnestFn(joinedSchema, unnestIndices)))
           .setRowSchema(joinedSchema);
     }
   }
 
   private static class UnnestFn extends DoFn<Row, Row> {
 
     private final Schema outputSchema;
-    private final int unnestIndex;
+    private final List<Integer> unnestIndices;
 
-    private UnnestFn(Schema outputSchema, int unnestIndex) {
+    private UnnestFn(Schema outputSchema, List<Integer> unnestIndices) {
       this.outputSchema = outputSchema;
-      this.unnestIndex = unnestIndex;
+      this.unnestIndices = unnestIndices;
     }
 
     @ProcessElement
     public void process(@Element Row row, OutputReceiver<Row> out) {
-
-      @Nullable Collection<Object> rawValues = row.getArray(unnestIndex);
+      Row rowWithArrayField = row;
+      Schema schemaWithArrayField = outputSchema;
+      for (int i = unnestIndices.size() - 1; i > 0; i--) {
+        rowWithArrayField = rowWithArrayField.getRow(unnestIndices.get(i));
```

Comment:
So it looks like guaranteed to have nested row except for the last index?

Reply:
I believe the updated code in `BeamUnnestRule.java` will guarantee this.